core/embed/sys/powerctl/npm1300/npm1300.c

```diff
@@ -0,0 +1,773 @@
+  drv->report_callback_context = context;
+
+  if (drv->report_callback != NULL) {
+    drv->adc_trigger_requested = true;
+    npm1300_fsm_continue(drv);
+  }
+
+  irq_unlock(irq_key);
+
+  return true;
+}
+
+// Synchronous measurement context structure
+// (used internally within the `npm1300_measure_sync` function)
+typedef struct {
+  // Set when the measurement is done
+  volatile bool done;
+  // Report structure where the measurement is stored
+  npm1300_report_t* report;
+} npm1300_sync_measure_t;
+
+// Callback for the synchronous measurement
+static void npm1300_sync_measure_callback(void* context,
+                                          npm1300_report_t* report) {
+  npm1300_sync_measure_t* ctx = (npm1300_sync_measure_t*)context;
+  *ctx->report = *report;
+  ctx->done = true;
+}
+
+bool npm1300_measure_sync(npm1300_report_t* report) {
+  npm1300_sync_measure_t measure = {
+      .done = false,
+      .report = report,
+  };
+
+  // Start asynchronous measurement
+  if (!npm1300_measure(npm1300_sync_measure_callback, &measure)) {
+    return false;
+  }
+
+  // Wait for the measurement to finish
+  while (!measure.done) {
+    __WFI();
+  }
+
+  return true;
+}
+
+// Prepares PMIC report from the last readout of the ADC values
+// stored in `drv->adc_regs`
+//
+// This function is called in the irq context.
+static void npm1300_calculate_report(npm1300_driver_t* drv,
+                                     npm1300_report_t* report) {
+  memset(report, 0, sizeof(npm1300_report_t));
+
+  npm1300_adc_regs_t* r = &drv->adc_regs;
+
+  // Gather measured values from the ADC registers
+
+  uint16_t vbat_adc =
+      (r->adc_vbat_result_msb << 2) + (r->adc_gp0_result_lsbs & 0x03);
+
+  uint16_t ntc_adc =
+      (r->adc_nt_result_msb << 2) + ((r->adc_gp0_result_lsbs >> 2) & 0x3);
+
+  uint16_t die_adc =
+      (r->adc_temp_result_msb << 2) + ((r->adc_gp0_result_lsbs >> 4) & 0x3);
+
+  uint16_t vsys_adc =
+      (r->adc_vsys_result_msb << 2) + ((r->adc_gp0_result_lsbs >> 6) & 0x03);
+
+  uint16_t ibat_adc =
+      (r->adc_vbat2_result_msb << 2) + ((r->adc_gp1_result_lsbs >> 4) & 0x03);
+
+  // IBAT_MEAS_STATUS register isn't well documented in the NPM1300 datasheet.
+  // The following is based partially on observation.
+  //
+  // 00100 - discharge
+  // 01000 - usb powered, not charging
+  // 01100 - charge trickle
+  // 01110 - charge cool
+  // 01111 - charge normal
+  // 1XXXX - invalid value, start measure again
+
+  // bool ibat_invalid = (ibat_meas_status & 0x10) != 0;
+  bool ibat_discharging = ((r->adc_ibat_meas_status >> 2) & 0x03) == 1;
+  bool ibat_charging = ((r->adc_ibat_meas_status >> 2) & 0x03) == 3;
+
+  if (ibat_discharging) {
+    report->ibat = ((int)ibat_adc * drv->i_limit) / 1250.0;
+  } else if (ibat_charging) {
+    report->ibat = -((int)ibat_adc * drv->i_charge) / 800.0;
+  } else {
+    report->ibat = 0;
+  }
+
+  report->vbat = (vbat_adc * 5.0) / 1023.0;
+  float beta = 3380;
+  report->ntc_temp =
```

Comment:
these conversions deserve its own functions/macros, to make this a bit more readable, the magic numbers should preferably be named

Reply:
Take a look at commit b5fb4de8d49521d1bde2c0a8684f428b9053ace9. I've added better comments explaining the code - these equations are all derived from the PMIC datasheet. I hope, it's more readable now.